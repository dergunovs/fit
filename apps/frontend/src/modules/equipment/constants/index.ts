export const URL_EQUIPMENT = '/equipment';
export const URL_EQUIPMENT_CREATE = `${URL_EQUIPMENT}/create`;
export const URL_EQUIPMENT_EDIT = `${URL_EQUIPMENT}/edit`;

export const EQUIPMENT_LIST_HEADERS = [{ title: 'Название' }];
