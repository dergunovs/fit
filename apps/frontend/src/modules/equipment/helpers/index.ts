import { IEquipment } from 'fitness-tracker-contracts';

export function filterEquipmentByWeights(equipments?: IEquipment[], isWeights?: boolean) {
  if (!equipments?.length) return [];

  return equipments?.filter((equipment) => (isWeights ? equipment.isWeights : !equipment.isWeights));
}
