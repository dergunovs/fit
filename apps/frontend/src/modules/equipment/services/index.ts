import { ComputedRef } from 'vue';
import {
  API_EQUIPMENT,
  TGetEquipmentsDTO,
  TGetEquipmentDTO,
  TPostEquipmentDTO,
  TPostEquipmentDataDTO,
  TUpdateEquipmentDTO,
  TUpdateEquipmentDataDTO,
  TDeleteEquipmentDTO,
} from 'fitness-tracker-contracts';
import { useMutation, useQuery, api } from 'mhz-helpers';

export const equipmentService = {
  getAll: () =>
    useQuery({
      queryKey: [API_EQUIPMENT],
      queryFn: async () => {
        const { data } = await api.get<TGetEquipmentsDTO>(API_EQUIPMENT);

        return data.data;
      },
    }),

  getOne: (options: object, id: ComputedRef<string>) =>
    useQuery({
      queryKey: [API_EQUIPMENT, id],
      queryFn: async () => {
        const { data } = await api.get<TGetEquipmentDTO>(`${API_EQUIPMENT}/${id.value}`);

        return data.data;
      },
      ...options,
    }),

  create: (options: object) =>
    useMutation({
      mutationKey: [API_EQUIPMENT],
      mutationFn: async (formData: TPostEquipmentDataDTO) => {
        const { data } = await api.post<TPostEquipmentDTO>(API_EQUIPMENT, formData);

        return data;
      },
      ...options,
    }),

  update: (options: object) =>
    useMutation({
      mutationKey: [API_EQUIPMENT],
      mutationFn: async (formData: TUpdateEquipmentDataDTO) => {
        const { data } = await api.patch<TUpdateEquipmentDTO>(`${API_EQUIPMENT}/${formData._id}`, formData);

        return data;
      },
      ...options,
    }),

  delete: (options: object) =>
    useMutation({
      mutationKey: [API_EQUIPMENT],
      mutationFn: async (id: string) => {
        const { data } = await api.delete<TDeleteEquipmentDTO>(`${API_EQUIPMENT}/${id}`);

        return data;
      },
      ...options,
    }),
};
