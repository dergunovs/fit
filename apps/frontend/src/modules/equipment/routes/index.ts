import { URL_EQUIPMENT, URL_EQUIPMENT_CREATE, URL_EQUIPMENT_EDIT } from '@/equipment/constants';

export const equipmentRoutes = [
  {
    path: URL_EQUIPMENT,
    name: 'EquipmentList',
    component: () => import('@/equipment/pages/EquipmentListPage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: URL_EQUIPMENT_CREATE,
    name: 'EquipmentCreate',
    component: () => import('@/equipment/pages/EquipmentCreatePage.vue'),
    meta: { layout: 'admin' },
  },
  {
    path: `${URL_EQUIPMENT_EDIT}/:equipment`,
    name: 'EquipmentEdit',
    component: () => import('@/equipment/pages/EquipmentEditPage.vue'),
    meta: { layout: 'admin' },
  },
];
