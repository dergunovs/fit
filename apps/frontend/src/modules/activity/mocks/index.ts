import { ref, computed } from 'vue';
import { vi } from 'vitest';
import { TDeleteActivityDTO, TGetActivitiesDTO } from 'fitness-tracker-contracts';

import * as activityComposables from '@/activity/composables';
import { activityService } from '@/activity/services';
import {
  ACTIVITIES_CALENDAR_FIXTURE,
  ACTIVITIES_FIXTURE,
  ACTIVITIES_STATISTICS_FIXTURE,
  ACTIVITY_FIXTURE,
  ACTIVITIES_CHART_FIXTURE,
} from '@/activity/fixtures';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { IOnSuccess } from '@/common/interface';

const dateFrom = ref('2024-11-24T21:00:00.000Z');
const dateTo = ref('2025-01-05T20:59:59.000Z');
const isDatesReady = ref(true);
const spyUpdateDates = vi.fn();

const spyGetActivity = vi.spyOn(activityService, 'getOne').mockReturnValue(mockQueryReply(ACTIVITY_FIXTURE));

const getActivitiesData: TGetActivitiesDTO = { data: ACTIVITIES_FIXTURE, total: ACTIVITIES_FIXTURE.length };

const spyGetActivities = vi
  .spyOn(activityService, 'getMany')
  .mockImplementation(() => mockQueryReply(getActivitiesData));

const spyDeleteActivity = vi.fn();

const onSuccess: IOnSuccess = {
  delete: undefined,
};

vi.spyOn(activityService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) onSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteActivityDTO, string>(spyDeleteActivity);
});

const spyUseActivityCalendar = vi.spyOn(activityComposables, 'useActivityCalendar').mockReturnValue({
  dateFrom,
  dateTo,
  isDatesReady,
  updateDates: spyUpdateDates,
});

const computedActivityCalendarEvents = computed(() =>
  activityComposables.convertActivityCalendarEvents(ACTIVITIES_CALENDAR_FIXTURE)
);

const spyComputedActivityCalendarEvents = vi
  .spyOn(activityComposables, 'computedActivityCalendarEvents')
  .mockImplementation(() => computedActivityCalendarEvents);

const spyGetActivitiesCalendar = vi
  .spyOn(activityService, 'getCalendar')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CALENDAR_FIXTURE));

const spyGetActivitiesStatistics = vi
  .spyOn(activityService, 'getStatistics')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_STATISTICS_FIXTURE));

const spyGetActivitiesChart = vi
  .spyOn(activityService, 'getChart')
  .mockImplementation(() => mockQueryReply(ACTIVITIES_CHART_FIXTURE));

export {
  spyGetActivities,
  spyGetActivity,
  spyGetActivitiesStatistics,
  spyGetActivitiesChart,
  spyDeleteActivity,
  spyUseActivityCalendar,
  spyComputedActivityCalendarEvents,
  spyGetActivitiesCalendar,
  spyUpdateDates,
  dateFrom,
  dateTo,
  isDatesReady,
  computedActivityCalendarEvents,
  getActivitiesData,
  onSuccess,
};
