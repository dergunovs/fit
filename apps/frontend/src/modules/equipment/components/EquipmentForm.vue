<template>
  <div>
    <UiFlex @submit.prevent="submit" tag="form" column gap="24" align="flex-start" data-test="equipment-form">
      <UiField :label="t('title')" isRequired :error="error('title')">
        <UiInput v-model="formData.title" data-test="equipment-form-title" />
      </UiField>

      <UiField :label="`${t('title')} EN`" :error="error('title_en')">
        <UiInput v-model="formData.title_en" data-test="equipment-form-title-en" />
      </UiField>

      <UiCheckbox
        v-model="formData.isWeights"
        :label="t('equipment.isWeights')"
        data-test="equipment-form-is-weights"
      />

      <FormButtons
        :id="props.equipment?._id"
        :isLoading="isLoadingPost || isLoadingUpdate"
        :isEdit="props.isEdit"
        @delete="(id) => mutateDelete(id)"
        data-test="equipment-form-buttons"
      />
    </UiFlex>
  </div>
</template>

<script setup lang="ts">
import { ref, onMounted } from 'vue';
import { useI18n } from 'vue-i18n';
import { useRouter } from 'vue-router';
import { UiCheckbox, UiField, UiFlex, UiInput, toast } from 'mhz-ui';
import { useQueryClient, useValidator, required, clone } from 'mhz-helpers';
import { API_EQUIPMENT, IEquipment } from 'fitness-tracker-contracts';

import FormButtons from '@/common/components/FormButtons.vue';

import { URL_EQUIPMENT } from '@/equipment/constants';
import { equipmentService } from '@/equipment/services';

interface IProps {
  equipment?: IEquipment | null;
  isEdit?: boolean;
}

const props = defineProps<IProps>();

const router = useRouter();
const { t, locale } = useI18n();
const queryClient = useQueryClient();

const formData = ref<IEquipment>({
  title: '',
  title_en: '',
  isWeights: false,
});

const { mutate: mutatePost, isPending: isLoadingPost } = equipmentService.create({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_EQUIPMENT] });
    toast.success(t('equipment.added'));
    router.push(URL_EQUIPMENT);
  },
});

const { mutate: mutateUpdate, isPending: isLoadingUpdate } = equipmentService.update({
  onSuccess: async () => {
    await queryClient.refetchQueries({ queryKey: [API_EQUIPMENT] });
    toast.success(t('equipment.updated'));
  },
});

const { mutate: mutateDelete } = equipmentService.delete({
  onSuccess: async () => {
    queryClient.removeQueries({ queryKey: [API_EQUIPMENT] });
    await queryClient.refetchQueries({ queryKey: [API_EQUIPMENT] });
    toast.success(t('equipment.deleted'));
    router.push(URL_EQUIPMENT);
  },
});

const { error, isValid } = useValidator(formData, {
  title: [required(locale.value)],
});

function submit() {
  if (!isValid()) return;

  if (props.equipment?._id) {
    mutateUpdate(formData.value);
  } else {
    mutatePost(formData.value);
  }
}

onMounted(() => {
  if (props.equipment) formData.value = clone(props.equipment);
});
</script>
