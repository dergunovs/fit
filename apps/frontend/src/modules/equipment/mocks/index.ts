import { vi } from 'vitest';
import {
  TDeleteEquipmentDTO,
  TPostEquipmentDataDTO,
  TPostEquipmentDTO,
  TUpdateEquipmentDataDTO,
  TUpdateEquipmentDTO,
} from 'fitness-tracker-contracts';

import { equipmentService } from '@/equipment/services';
import { mockMutationReply, mockQueryReply } from '@/common/mocks';
import { EQUIPMENT_FIXTURE, EQUIPMENTS_FIXTURE } from '@/equipment/fixtures';
import { IOnSuccess } from '@/common/interface';

const spyGetEquipment = vi.spyOn(equipmentService, 'getOne').mockReturnValue(mockQueryReply(EQUIPMENT_FIXTURE));

const spyGetEquipments = vi
  .spyOn(equipmentService, 'getAll')
  .mockImplementation(() => mockQueryReply(EQUIPMENTS_FIXTURE));

const spyCreateEquipment = vi.fn();
const spyUpdateEquipment = vi.fn();
const spyDeleteEquipment = vi.fn();

const mockOnSuccess: IOnSuccess = {
  create: undefined,
  update: undefined,
  delete: undefined,
};

vi.spyOn(equipmentService, 'create').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.create = options.onSuccess;

  return mockMutationReply<TPostEquipmentDTO, TPostEquipmentDataDTO>(spyCreateEquipment);
});

vi.spyOn(equipmentService, 'update').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.update = options.onSuccess;

  return mockMutationReply<TUpdateEquipmentDTO, TUpdateEquipmentDataDTO>(spyUpdateEquipment);
});

vi.spyOn(equipmentService, 'delete').mockImplementation((options: { onSuccess?: () => Promise<void> }) => {
  if (options.onSuccess) mockOnSuccess.delete = options.onSuccess;

  return mockMutationReply<TDeleteEquipmentDTO, string>(spyDeleteEquipment);
});

export { spyGetEquipment, spyGetEquipments, spyCreateEquipment, spyUpdateEquipment, spyDeleteEquipment, mockOnSuccess };
