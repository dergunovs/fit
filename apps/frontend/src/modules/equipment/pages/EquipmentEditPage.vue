<template>
  <div>
    <EquipmentForm v-if="equipment" :equipment="equipment" data-test="equipment-form" />
  </div>
</template>

<script setup lang="ts">
import { useRouteId } from 'mhz-helpers';

import EquipmentForm from '@/equipment/components/EquipmentForm.vue';

import { equipmentService } from '@/equipment/services';

const { id } = useRouteId('equipment');

const { data: equipment } = equipmentService.getOne({}, id);
</script>
