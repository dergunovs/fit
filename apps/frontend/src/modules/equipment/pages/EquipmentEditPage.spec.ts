import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import EquipmentEditPage from './EquipmentEditPage.vue';
import EquipmentForm from '@/equipment/components/EquipmentForm.vue';

import { wrapperFactory } from '@/common/test';
import { spyGetEquipment } from '@/equipment/mocks';
import { EQUIPMENT_FIXTURE } from '@/equipment/fixtures';
import { spyUseRouteId, mockRouteId } from '@/common/mocks';

const equipmentForm = dataTest('equipment-form');

let wrapper: VueWrapper<InstanceType<typeof EquipmentEditPage>>;

beforeEach(() => {
  wrapper = wrapperFactory(EquipmentEditPage);
});

enableAutoUnmount(afterEach);

describe('EquipmentEditPage', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(EquipmentEditPage)).toBeTruthy();
  });

  it('gets and sets equipment to form', async () => {
    expect(spyUseRouteId).toBeCalledTimes(1);
    expect(spyUseRouteId).toBeCalledWith('equipment');

    expect(spyGetEquipment).toBeCalledTimes(1);
    expect(spyGetEquipment).toBeCalledWith({}, mockRouteId);

    expect(wrapper.findComponent<typeof EquipmentForm>(equipmentForm).vm.$props.equipment).toStrictEqual(
      EQUIPMENT_FIXTURE
    );
  });
});
