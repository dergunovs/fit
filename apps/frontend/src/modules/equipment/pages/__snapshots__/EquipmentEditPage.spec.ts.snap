// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EquipmentEditPage > matches snapshot 1`] = `
"<div>
  <equipment-form-stub equipment="[object Object]" isedit="true" data-test="equipment-form"></equipment-form-stub>
</div>"
`;
